### ISSUE TITLE
Support for multi-file scripts

### ISSUE THREAD

**Issue by CyberShadow** [koalaman/shellcheck#913]
Currently, if you have a script which consists of multiple files (e.g. `main.sh` which sources `lib/a.sh`, `lib/b.sh` etc.), there is no straight-forward way to validate them using ShellCheck. Although the `-x` option allows ShellCheck to read sourced scripts, it will not display any warnings in them.

Some current options are:

1. Preprocess and concatenate all scripts into a single file. This also requires replacing/removing the `source` directives (or excluding "not sourced" warnings).

   The main problem with this approach is that file names and line numbers will no longer correspond to the actual source code.

2. Use fake `source` statements in each file, e.g. `lib/b.sh` could contain:

   ```bash
   # shellcheck source=lib/a.sh
   source /dev/null
   ```

   (see #905)

   The problem with this approach is that as soon as you add such directives to a second file, you will have infinite loops, or things will not be evaluated in the correct order. (`a.sh` may need to be checked together with `c.sh` to satisfy "variable is declared but not used" type warnings; therefore, "sourcing" `a.sh` from `b.sh` will cause ShellCheck to evaluate `c.sh` before the body of `b.sh`).

3. For each file, write a mock "header" script which declares variables actually declared in other source files. E.g. `b.sh` could have a companion `b.env` which contains stub declarations for variables/functions actually declared in `a.sh`.

Suffice to say, none of these workarounds are satisfactory.

One way to solve this problem is to add a switch to `shellcheck`, e.g. `-r`, which instructs ShellCheck to not just read, but also validate (show warnings in) sourced files. 

A WIP for implementing this is here: https://github.com/koalaman/shellcheck/compare/master...CyberShadow:recursive

Some implementation challenges are:

- The program must know that it needs to display not just errors, but also warnings in sourced files.
- The output formatters currently seem to expect that all errors will be encountered in a single file. (At least, when disabling filtering of error messages in other files, the displayed filename is wrong, despite the program recording the file name in the context being read.)
- It may be useful to limit errors from which files are to be displayed.

**Comment by CyberShadow** [comment]
> Preprocess and concatenate all scripts into a single file.

I have now written a tool which does this, and translates file names / line numbers back to their original one:

https://github.com/CyberShadow/multishellcheck

**Comment by koalaman** [comment]
I'm not surprised you went that route. This was not nearly as straight forward as I imagined.

* Formatters didn't support multiple files, like you said
* The formatters expected the contents of *the* script. There can be multiple scripts, and you can 't just re-read stdin when you need it.
* Threading this option through both parsing and analysis required some broad if shallow changes.

8dd40ef finally adds support for this through a `-a` though.
